nasa/ow_autonomy
# Expose and utilize pan and tilt faults in PLEXIL

## Linked Issues:
| EPIC ⚡ | N/A |
| :----------- | :----------- |
| Jira Ticket 🎟️ | [OCEANWATER-779](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-779) |
| Github :octocat:  | # |


## Summary of Changes
* New Lookups `AntennaPanFault` and `AntennaTiltFault` in addition to the existing `AntennaFault`.
* New  test plan `TestFaults.plp`

## Test
* Fetch the `OCEANWATER-934_telemetry_unification_epic` branch of `ow_simulator` to accompany this branch.
* Clean build, resource setup files.
* Start any simulator world.
* In other terminal, run `roslaunch ow_plexil ow_exec.launch`
* Run the plan `TestFaults.plx`
* Inject and clear antenna pan and tilt joint locked faults in any order, and see that the ROS messages printed in the ow_plexil terminal are accurate.  An example sequence is as follows.

```
[ INFO]: PLEXIL: Inject faults and observe messages. Type Control-C to exit plan and exec node.
[ INFO]: PLEXIL: Antenna is fine.
[ INFO]: PLEXIL: Pan is fine.
[ INFO]: PLEXIL: Tilt is fine.
[ INFO]: PLEXIL: Antenna is fine.
[ INFO]: PLEXIL: Pan is fine.
[ INFO]: PLEXIL: Tilt is fine.
[ INFO]: PLEXIL: Antenna is fine.
[ INFO]: PLEXIL: Pan is fine.
[ INFO]: PLEXIL: Tilt is fine.
[ WARN]: Fault in ANTENNA: TiltFault
[ WARN]: Fault in SYSTEM: PAN_TILT_EXECUTION_ERROR
[ WARN]: PLEXIL: Antenna fault present.
[ INFO]: PLEXIL: Pan is fine.
[ WARN]: PLEXIL: Tilt fault present.
[ WARN]: PLEXIL: Antenna fault present.
[ INFO]: PLEXIL: Pan is fine.
[ WARN]: PLEXIL: Tilt fault present.
[ WARN]: PLEXIL: Antenna fault present.
[ INFO]: PLEXIL: Pan is fine.
[ WARN]: PLEXIL: Tilt fault present.
[ WARN]: Fault in ANTENNA: PanFault
[ WARN]: PLEXIL: Antenna fault present.
[ WARN]: PLEXIL: Pan fault present.
[ WARN]: PLEXIL: Tilt fault present.
[ WARN]: Resolved fault in ANTENNA: TiltFault
[ WARN]: PLEXIL: Antenna fault present.
[ WARN]: PLEXIL: Pan fault present.
[ INFO]: PLEXIL: Tilt is fine.
[ WARN]: PLEXIL: Antenna fault present.
[ WARN]: PLEXIL: Pan fault present.
[ INFO]: PLEXIL: Tilt is fine.
[ WARN]: PLEXIL: Antenna fault present.
[ WARN]: PLEXIL: Pan fault present.
[ INFO]: PLEXIL: Tilt is fine.
[ WARN]: PLEXIL: Antenna fault present.
[ WARN]: PLEXIL: Pan fault present.
[ INFO]: PLEXIL: Tilt is fine.
[ WARN]: Resolved fault in ANTENNA: PanFault
[ WARN]: Resolved fault in SYSTEM: PAN_TILT_EXECUTION_ERROR
[ INFO]: PLEXIL: Antenna is fine.
```
### Notes

- This is a scaled down version of @pthangeda's original work, which at present is not achievable due to the next point.
- Even though pan and tilt faults can now be detected separately, this is of limited use because Pan and Tilt can no longer be commanded separately as they were before.  We plan to bring back this feature, at which time @pthangeda's fault handling example plan can be restored.
